Refactor the Permission checking decorators

Acidity opened the pull request.

Acidity committed:
Separate the user permission check decorators into 2 decorators.

This removes the accept_any_matching argument from user_has_permission,
and moves that functionality to a new decorator called
user_has_any_permission.

Signed-off-by: Tyler O'Meara <Tyler@TylerOMeara.com>

Acidity committed:
Move the runtime permission handling code into it's own decorator.

Signed-off-by: Tyler O'Meara <Tyler@TylerOMeara.com>

jboning commented on brave/core/permission/util.py:
```diff
@@ -11,39 +11,18 @@
-        accept_any_matching=True, but not when wild=False."""
-        
+def prepare_runtime_permission(perm=None, **runkw):
+    """This decorator handles the runtime permission aspects of the user permission checking decorators."""
```
I was pretty confused at first about what this actually did. Is there a reason you wrote it as a decorator rather than a normal helper function?

jboning commented:
I like the user_has_permission/user_has_any_permission split!

Acidity committed:
Make prepare_runtime_permission a function.

Also fix a bug when evaluating if a user has any permission where
if they had the wildcard permission they could incorrectly be marked
as not having the permission.

Signed-off-by: Tyler O'Meara <Tyler@TylerOMeara.com>

jboning commented on brave/core/account/model.py:
```diff
@@ -193,6 +193,8 @@ def has_any_permission(self, permission):
             if p.grants_permission(permID.id):
                 return True
+            if permID.grants_permission(p.id):
+                return True
```
What's going on here?

Acidity replied:
This fixes a couple of bugs, mainly centered around the fact that if we're checking for any permission of the form 'core.group.*', then the user having 'core.*' should pass evaluation.

jboning commented on brave/core/permission/util.py:
```diff
@@ -11,27 +11,47 @@
         
             user = user._current_obj()
+
+            permission = prepare_runtime_permission(self, perm, runkw, args, kwargs)
```
check for False here?

jboning commented on brave/core/permission/util.py:
```diff
@@ -42,34 +62,53 @@ def check_permission(self, *args, **kwargs):
+        
+            user = user._current_obj()
+
+            permission = prepare_runtime_permission(self, perm, runkw, args, kwargs)
```
same question

jboning commented on brave/core/permission/util.py:
```diff
@@ -42,34 +62,53 @@ def check_permission(self, *args, **kwargs):
+            for permID in user.permissions:
+                if p.grants_permission(permID.id):
+                    return function(self, *args, **kwargs)
+                if permID.grants_permission(p.id):
```
same question as the first one, this seems to be a behaviour change?

Acidity committed:
Fix permission decorators when a permission is not provided.

Signed-off-by: Tyler O'Meara <Tyler@TylerOMeara.com>
